tests/s25Main/integration/testPacts.cpp

```diff
@@ -261,4 +261,196 @@ BOOST_FIXTURE_TEST_CASE(SuggestNewPactAfterExpiration, PactCreatedFixture,
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == false);
+
+    p1.OnAttackedBy(0);
+    BOOST_TEST_REQUIRE(p0.IsAlly(1) == false);
+    BOOST_TEST_REQUIRE(p1.IsAlly(0) == false);
+    BOOST_TEST_REQUIRE(p0.CanAttack(1) == true);
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == true);
+}
+
+using EmptyWorldFixture3P = WorldFixture<CreateEmptyWorld, 3>;
+
+BOOST_FIXTURE_TEST_CASE(
+  OneSidedAlliances_BeingAttacked_AlsoBreaksAlliancesForPlayers_WhoAreAlliedToUs_AndWhoWeAreAlliedTo,
+  EmptyWorldFixture3P)
+{
+    auto& p0 = world.GetPlayer(0);
+    auto& p1 = world.GetPlayer(1);
+    auto& p2 = world.GetPlayer(2);
+    p0.ps = PlayerState::AI;
+    p1.ps = PlayerState::AI;
+    p2.ps = PlayerState::AI;
+
+    p0.MakeOneSidedAllianceTo(1);
+    p0.MakeOneSidedAllianceTo(2);
+    p1.MakeOneSidedAllianceTo(0);
+    p1.MakeOneSidedAllianceTo(2);
+    p2.MakeOneSidedAllianceTo(0);
+
+    BOOST_TEST_REQUIRE(p0.CanAttack(1) == false);
+    BOOST_TEST_REQUIRE(p0.CanAttack(2) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(2) == false);
+    BOOST_TEST_REQUIRE(p2.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p2.CanAttack(1) == true);
+
+    p1.OnAttackedBy(2);
+
+    // alliance from p1 to p2 is broken
+    // but alliance between p0 and p2 is broken as well
+    BOOST_TEST_REQUIRE(p0.CanAttack(1) == false);
+    BOOST_TEST_REQUIRE(p0.CanAttack(2) == true);
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(2) == true);
+    BOOST_TEST_REQUIRE(p2.CanAttack(0) == true);
+    BOOST_TEST_REQUIRE(p2.CanAttack(1) == true);
+}
+
+BOOST_FIXTURE_TEST_CASE(OneSidedAlliances_BeingAttacked_DoesNotAlsoBreakAlliancesForPlayers_WhoAreNotOurAlliedToUs,
+                        EmptyWorldFixture3P)
+{
+    auto& p0 = world.GetPlayer(0);
+    auto& p1 = world.GetPlayer(1);
+    auto& p2 = world.GetPlayer(2);
+    p0.ps = PlayerState::AI;
+    p1.ps = PlayerState::AI;
+    p2.ps = PlayerState::AI;
+
+    p0.MakeOneSidedAllianceTo(2);
+    p1.MakeOneSidedAllianceTo(0);
+    p1.MakeOneSidedAllianceTo(2);
+    p2.MakeOneSidedAllianceTo(0);
+
+    BOOST_TEST_REQUIRE(p0.CanAttack(1) == true);
+    BOOST_TEST_REQUIRE(p0.CanAttack(2) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(2) == false);
+    BOOST_TEST_REQUIRE(p2.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p2.CanAttack(1) == true);
+
+    p1.OnAttackedBy(2);
+
+    // alliance from p1 to p2 is broken
+    // alliance from p0 to p2 remains
+    BOOST_TEST_REQUIRE(p0.CanAttack(1) == true);
+    BOOST_TEST_REQUIRE(p0.CanAttack(2) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p1.CanAttack(2) == true);
+    BOOST_TEST_REQUIRE(p2.CanAttack(0) == false);
+    BOOST_TEST_REQUIRE(p2.CanAttack(1) == true);
+}
+
+BOOST_FIXTURE_TEST_CASE(OneSidedAlliances_BeingAttacked_DoesNotAlsoBreakAlliancesForPlayers_WhoWeAreNotAlliedTo,
```

Comment:
I guess this needs resolving the definition of "allied to".

Say player X attacks player Y. Then of course player Y breaks the pact to player X.

And I guess every player Z allied to player Y should also break the pact to player X. Only question is whether this Z->Y should include one-sided pacts and if so which way: Z->Y, Y->Z, both or either. I would argue only Z->Y should be checked as Z "doesn't know" about Y->Z so why should it care about Y being attacked in that case?

Might also be worth merging with the above test case to focus on the differences and add the "DoesNotAlsoBreakAlliancesForPlayers_WhoWeAreNotAlliedTo" as a comment instead